torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,202 @@
+                for n in lookup.keys():
+                    if n.op == "placeholder" or lookup[n].op == "output":
+                        continue
+                    for user in n.users:
+                        # If this node has users that were not in
+                        # `lookup`, then it must leak out of the
+                        # pattern subgraph
+                        if user not in lookup:
+                            return False
+                return True
+
+            # It's not a match if the pattern leaks out into the rest
+            # of the graph
+            if pattern_is_contained(matcher.nodes_map):
+                # Shallow copy nodes_map
+                matches.append(Match(anchor=anchor,
+                                     nodes_map=copy(matcher.nodes_map)))
+
+    # The set of all nodes in `original_graph` that we've seen thus far
+    # as part of a pattern match
+    nodes_to_delete: Set[Node] = set()
+
+    # Get the Nodes we'll use to hook the replacement subgraph in to the
+    # original graph
+    replacement_inputs: List[Node] = [n for n in replacement_graph.nodes
+                                      if n.op == "placeholder"]
+    replacement_output: Node = next(n for n in replacement_graph.nodes
+                                    if n.op == "output")
+
+    # Return TRUE if one of the nodes in the current match has already
+    # been used as part of another match
+    def overlaps_with_prev_match(match: Match) -> bool:
+        for n in match.nodes_map.values():
+            if n in nodes_to_delete:
+                return True
+        return False
+
+    if not matches:
+        raise ValueError("No matches were found for the given pattern")
+
+    for match in matches:
+
+        # Skip overlapping matches
+        if overlaps_with_prev_match(match):
+            continue
+
+        # Get the original graph nodes corresponding to the "start" and
+        # "end" of the pattern subgraph. These connections are where
+        # we'll hook the replacement subgraph in to the original graph
+        original_graph_inputs: Dict[str, Node] = {n.name : n
+                                                  for n
+                                                  in original_graph.nodes
+                                                  if n.op == "placeholder"}
+        subgraph_output: Node = match.anchor
+
+        def mark_nodes_for_deletion(n: Node) -> None:
+            if n not in match.nodes_map.values():
+                return
+            for n_ in n.uses:
+                mark_nodes_for_deletion(n_)
+            nodes_to_delete.add(n)
+            # By removing the `original_graph` node's name early, we
+            # ensure that we can insert nodes from `replacement` with
+            # the same name. (This is strictly cosmetic)
+            if n.name in original_graph._used_names:
+                original_graph._used_names.pop(n.name)
```

Comment:
`_used_names` maps from base name to the highest index that's been used for SSAification (e.g. name 'foo' could appear as 'foo' -> 1 if there's a name 'foo_1'), so removing the name from `_used_names` probably isn't correct. Can we just let the names get mangled normally here?